Average GPA fix

Edwu29 opened the pull request:
## Reference Issues
Addresses Mr. Zotistics found issue with the average of the 'averageGPA' field. The average GPA included P/NP fields where the 'averageGPA' was empty which counts as 0 towards the total average. More info: 

> Alight, so there is something wrong with the GPA calculations. I think both the API and current Zotistics calculations are wrong and I think it's cuz the gpa average that the school gave us in the data is different than how you would normally think to calculate it.
> 
> I'll use Pattis as an example:
> If you use the raw numbers (ignoring P/NP), you'd get this
> ((3998*4)+(2402*3)+(1339*2)+(470))/(3998+2402+1339+470+834) = 2.9134
> 
> which is different than what shows up on Zotistics: 3.073 and on the API: 2.70
> 
> How I calculated that on Zotistics is get the mean of the gpa average column (it disregards the P/NP classes because the avg is NaN)
> round(self.searchDF['avg'].mean(), 3)
> 
> On the React version of Zotistics, I just use what's given to me on the API (unless an advanced option is used).
> 
> So I assume when getting the the average GPA in the SQL query for the API
> AVG(averageGPA)
> 
> it's including the P/NP as 0 gpa which is why the gpa is significantly lower.
> 
> But even if corrected, idk if the result (the same as Flask Zotistics) is actually even correct.

## Summary of Change/Fix 
### REST API

- Added another query parameter `'excludePNP'`.
- Changed the helper function `queryDatabaseAndResponse` in helper/grades.helper.js to add the condition `AND averageGPA != ''` in the SQL statement.

### GraphQL API

- Updated the GraphQL query for grades to include `'excludePNP'` argument of `GraphQLBoolean` type. 
- Changed the helper function `fetchAggregatedGrades` in helper/grades.helper.js to add the condition in the SQL statement, same as REST.

### Documentation

- Updated documentation to add 'excludePNP' parameter for REST API and added another query example for GraphQL API under `GraphQL-API/start_here`
![image](https://user-images.githubusercontent.com/39328613/132074113-479193c4-db85-43de-973d-ab9d90ede3f6.png)


## Test Plan

Checked the contents of the `db.sqlite` to look at the AVG(averageGPA) of including/excluding where averageGPA is empty.
Using sqlite VSCode extension to look at the table. 

Using Pattis as example: 
Including P/NP, there are 86 rows:
![image](https://user-images.githubusercontent.com/39328613/132072488-07318274-6fc2-4a91-b025-cb3f8a8bd9e7.png)

Excluding P/NP, there are 76 rows, increasing the AVG of averageGPA field:
![image](https://user-images.githubusercontent.com/39328613/132072470-0405b995-5170-4686-afaf-f9d703368b1d.png)

This is further verified by viewing the table to make sure empty averageGPA tuples were removed. 

### REST API
Tested using the same Pattis example to get the same AVG of averageGPA field, Including P/NP:
![image](https://user-images.githubusercontent.com/39328613/132072163-2da9663e-2581-439d-b41c-5a2ed4bbe755.png)

Excluding P/NP:
![image](https://user-images.githubusercontent.com/39328613/132072222-c1f7d7e5-27b3-4ecd-b68b-a058955a51b8.png)

### GraphQL API
Tested using GraphQL playground using the same Pattis example, Including P/NP:
![image](https://user-images.githubusercontent.com/39328613/132072369-67d06d77-e9ea-4086-bb4d-3753b8a6000f.png)

Excluding P/NP:
![image](https://user-images.githubusercontent.com/39328613/132072408-48f851a1-0d21-4645-97fe-f1341acc361b.png)

ChaseC99 commented:
Really detailed PR - great work! I'll review it fully later tonight. 

Just an FYI, it's currently failing some of the unit tests. You may want to take a look at that: https://github.com/icssc-projects/peterportal-public-api/pull/102/checks?check_run_id=3510706352

Edwu29 commented:
I touched some round brackets which caused the unit tests to fail. Tests passed now!

ramanxg commented on docs/GraphQL-API/start_here.md:
````diff
@@ -26,6 +26,26 @@ Not only can you try it out, but our Playground also has the most up-to-date doc
     }
     ```
 
+!!! example "Example Query 2"
````
I'm not seeing the built docs. Since we currently don't have automated docs building, can you run `mkdocs build -d docs-site`. You can check by checking the documentation running `npm start` and seeing that your new changes are there.

Edwu29 replied:
Gotcha

ramanxg reviewed (commented):
Awesome stuff. :) Left a few comments that you can take a look at

ramanxg commented on helpers/grades.helper.js:
```diff
@@ -102,30 +103,38 @@ function fetchInstructors(where) {
+function queryDatabaseAndResponse(where, calculate, excludePNP = 'true') {
     const connection = new db(path.join(__dirname, '../db/db.sqlite'));
-
+    const boolExcludePNP = (excludePNP.toLowerCase() == 'true') //passOrNoPass string->bool
```
Think it would be better to convert it to a boolean in `/rest/v0/grades.js`. Since it's a paramter from the user, we should handle it in the endpoint function. This way, this function also becomes similar to the one used for graphql.

Edwu29 replied:
Makes sense!

ramanxg commented on helpers/grades.helper.js:
```diff
@@ -102,30 +103,38 @@ function fetchInstructors(where) {
 
-function fetchAggregatedGrades(where) {
+//For GraphQL API
+function fetchAggregatedGrades(where, excludePNP = true) {
```
I think I like having it default to `false`. I would rather for a default request, to return all the classes that match criteria. We include as much data as possible and exclude data only when specified.

Edwu29 replied:
Agreed

Edwu29 replied:
True or false should be chosen in the endpoint, but I added a default param in `parseGradesParamsToSQL` for GraphQL, although it seems to be false on default regardless, not sure how.

Edwu29 commented on helpers/grades.helper.js:
```diff
@@ -102,30 +103,38 @@ function fetchInstructors(where) {
+    
+    if (where !== null){
+        sqlStatement = sqlStatement + where
+        if(excludePNP){
+            sqlStatement = sqlStatement + ` AND (averageGPA != '')` //make sure to use single quote '' instead of "" in SQLite
+        }
```
Agreed

Edwu29 replied:
I tried passing `excludePNP` within `query` in `parseGradesParamsToSQL` but ran into the issue of it being stored as a string so I passed it as another param instead. Also tried resolving the condition within the switch statement in `parseGradesParamsToSQL` but also ran into issues, so I added the logic at the end. I believe this works best, but let me know if there's a cleaner way.

Edwu29 commented:
Test Plan passes. Issues fixed, left some comments regarding issues.

ramanxg commented on helpers/grades.helper.js:
```diff
@@ -4,7 +4,8 @@ var path = require('path');
 
-function parseGradesParamsToSQL(query) {
+// Constructs a WHERE clause from the query
+function parseGradesParamsToSQL(query, excludePNP = false) {
```
It seems like for both graphql and rest, excludePNP is originally part of the `query` object. What do you think about removing the extra argument, and utilizing `query.excludePNP`?

Edwu29 replied:
Agreed!

Edwu29 commented:
## Summary of Change/Fix 

- Changed the `average_gpa` field to always exclude empty `averageGPA` tuples as requested and @ramanxg for the clever fix using NULLIF in the following fashion: `AVG(NULLIF(averageGPA, ''))`
- Updated tests to reflect changes in field names
- added `count` field to GraphQL API 
## Updated Test Plan

### REST API
excludePNP = false:
![image](https://user-images.githubusercontent.com/39328613/133159105-34268365-cb75-4315-ae6a-3a357bfaa32a.png)
excludePNP = true:
![image](https://user-images.githubusercontent.com/39328613/133159198-9ccddab9-8692-46fa-b110-67f268ce790e.png)
average_gpa now remains the same regardless of `excludePNP` param. The `count`,` sum_grade_p_count`, `sum_grade_np_count`, `sum_grade_w_count` varies by excluding/including PNP.
### GraphQL API
excludePNP = false:
![image](https://user-images.githubusercontent.com/39328613/133159336-4b19885c-dd34-4ddb-a8c6-29c574de9e49.png)
excludePNP = true:
![image](https://user-images.githubusercontent.com/39328613/133159378-29e2c6d0-ba60-493e-8fbb-db2489daa408.png)
Same results are reflected through GraphQL query.

ramanxg commented on rest/v0/grades.js:
```diff
@@ -23,9 +23,11 @@ router.get("/raw", async (req, res) => {
+        req.query.excludePNP = (req.query.excludePNP == 'true') ? true : false // converting excludePNP string -> bool
         const where = parseGradesParamsToSQL(req.query);
-        const results = queryDatabaseAndResponse(where, true)
+        const results = queryDatabaseAndResponse(where, true) //excludePNP adds another parameter to the query
```
We don't need this comment, since we don't use the parameter here anymore.

Edwu29 replied:
Yup

ramanxg commented on docs/REST-API/grades.md:
```diff
@@ -107,6 +107,7 @@ Try out one of these quick ways to play with our grades endpoints:
 |  `code` | 5-digit course code on WebSoC | 35540 |
+| `excludePNP` | Exclude P/NP | true/false |
 
 
```
In this doc, can we also change the naming of the /grades/calculated, so it matches to the new names you changed.

Edwu29 replied:
Agreed

ramanxg reviewed (changes requested):
This looks good to me, some small things, after they are fixed I think we can merge unless @ChaseC99 has anything else to add.
